### PR TITLE
feat: improve code quality

### PR DESCRIPTION
**Pull request by k-g-j**
## Description

Improves code quality by:
* Moving validation logic for ChainConfig into util
* Dynamically generates expected salt address in create account test
* Removes localhost:8080 from allowedOrigins
* Adds JSDoc comments to public keyring methods

 Fixes [#256](https://app.zenhub.com/workspaces/metamask-accounts-team-v2-64c91cbeaa9d1c00126621fd/issues/gh/metamask/accounts-planning/256)
See: [#244](https://app.zenhub.com/workspaces/metamask-accounts-team-v2-64c91cbeaa9d1c00126621fd/issues/gh/metamask/accounts-planning/244)

### CODE REVIEW

File: packages/snap/src/utils/validation.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    config.customVerifyingPaymasterAddress,
+    'Custom Verifying Paymaster Address',
+  );
+  validateUrl(config.bundlerUrl, 'Bundler URL');
+  validatePrivateKey(
+    config.customVerifyingPaymasterPK,
+    'Custom Verifying Paymaster Private Key',
+  );
+}
+
```
Comment (montelaidev) [inline packages/snap/src/utils/validation.ts:28]: I think we can use super struct here

Comment (k-g-j) [reply packages/snap/src/utils/validation.ts]: Done!

File: packages/snap/src/utils/validation.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import { ethers } from 'ethers';
+import type { ChainConfig } from 'src/keyring';
+import { assert, define, object, optional, StructError } from 'superstruct';
+
+import { throwError } from './util';
+
+const EthereumAddress = define(
+  'EthereumAddress',
+  (value) => typeof value === 'string' && ethers.isAddress(value),
+);
```
Comment (gantunesr) [inline packages/snap/src/utils/validation.ts:10]: Good use of `define` to great data validators! As a recommendation I suggest to take a look at the module `@metamask/utils` where we already developed a lot of useful validators like [`isValidHexAddress`](https://github.com/MetaMask/utils/blob/e97f2460d23ec9ed0907e5b68af8ecb84b28cb5a/src/hex.ts#L76)

File: packages/snap/src/utils/validation.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+const Url = define('Url', (value) => {
+  const urlPattern =
+    /^(https?:\/\/)?[\w\\.-]+(:\d{2,6})?(\/[\\/\w \\.-]*)?(\?[\\/\w .\-=]*)?$/u;
```
Comment (gantunesr) [inline packages/snap/src/utils/validation.ts:14]: I see in the regex that you're considering backslashes and spaces (`[\w\\.-]+` and `(\/[\\/\w \\.-]*)?`), these are quite unusual for urls. Are you validating a specific case?

Also, take into consideration that [in urls spaces are encoded as `%20`](https://developers.google.com/maps/url-encoding#:~:text=For%20example%2C%20spaces%20in%20a,should%20be%20encoded%20as%20%2C%20.) so I'm wondering if having it in the regex is necessary.

File: packages/snap/src/keyring.ts [hunk 4]
```diff
@@ -146,17 +115,37 @@ export class AccountAbstractionKeyring implements Keyring {
...
+   */
   async getAccount(id: string): Promise<KeyringAccount> {
     return (
       this.#state.wallets[id]?.account ??
       throwError(`Account '${id}' not found`)
     );
   }
 
+  /**
+   * Create a new account abstraction keyring account.
```
Comment (gantunesr) [inline packages/snap/src/keyring.ts:142]: Is it right to call it `account abstraction keyring account`? Maybe `smart contract account` is more accurate?

File: packages/snap/src/utils/validation.ts (hunk 3 above)
Comment (k-g-j) [reply packages/snap/src/utils/validation.ts:17]: updated urlPattern
```ts
  const urlPattern =
    /^(https?:\/\/)?[\w.-]+(:\d{2,6})?(\/[\w\/.-]*)?(\?[\w\/.&=\-%]*)?$/u;
    ```
    * Removed the match for backslashes (\\) and spaces (now encoded as %20 or +)
    * Added the percent character (%) to both the path and query string parts

File: packages/snap/src/utils/validation.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    /^(https?:\/\/)?[\w.-]+(:\d{2,6})?(\/[\w.-]*)?(\?[\w.&=%-]*)?$/u;
+  return typeof value === 'string' && urlPattern.test(value);
+});
+
+const PrivateKey = define('PrivateKey', (value) => {
+  return typeof value === 'string' && ethers.isHexString(value);
+});
+
+const ChainConfigStruct = object({
+  [CONFIG_KEYS.SIMPLE_ACCOUNT_FACTORY]: optional(EthereumAddress),
```
Comment (montelaidev) [inline packages/snap/src/utils/validation.ts:26]: We should use `exactOptional` here from `keyring-api`, using optional can lead to the failure of the state to be saved in the case were it is undefined.

File: packages/snap/src/keyring.test.ts [hunk 6]
```diff
@@ -104,14 +104,14 @@ describe('Keyring', () => {
...
+      simpleAccountFactory: '0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0',
+      entryPoint: '0x5FbDB2315678afecb367f032d93F642f64180aa3',
       bundlerUrl: 'https://bundler.example.com/rpc',
       customVerifyingPaymasterPK: aaOwnerPk,
       customVerifyingPaymasterAddress:
-        '0x123456789ABCDEF0123456789ABCDEF012345678',
+        '0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512',
     };
-    it('should not set the config without a valid config', async () => {
+    it('should not set the config with and invalid simpleAccountFactory address', async () => {
```
Comment (montelaidev) [inline packages/snap/src/keyring.test.ts:114]:
```suggestion
  it('should not set the config with an invalid simpleAccountFactory address'
```

Nit: typo here

File: packages/snap/src/utils/validation.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  try {
+    assert(config, ChainConfigStruct);
+  } catch (error) {
+    if (error instanceof StructError) {
+      let customMessage = `[Snap] Invalid chain configuration: ${error.message}`;
+      const { path, value } = error;
+      if (path.length > 0) {
+        const fieldName = path[0];
+        switch (fieldName) {
+          case 'simpleAccountFactory':
```
[the remark this answers is not in the document]
Comment (k-g-j) [reply packages/snap/src/utils/validation.ts]: Addressed here [ba7c204](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/ba7c204543d5d237dbf62413cf4cf97985a2bc53)

File: packages/snap/src/utils/validation.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * @throws If the configuration is invalid.
+ */
+export function validateConfig(config: ChainConfig): void {
+  try {
+    assert(config, ChainConfigStruct);
+  } catch (error) {
+    if (error instanceof StructError) {
+      let customMessage = `[Snap] Invalid chain configuration: ${error.message}`;
+      const { path, value } = error;
+      if (path.length > 0) {
```
[the remark this answers is not in the document]
Comment (k-g-j) [reply packages/snap/src/utils/validation.ts]: Addressed here [d2303e6](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/d2303e639c46dad9cf82e1104e0a71756ca02d79)

File: packages/snap/src/utils/validation.ts (hunk 2 above)
Comment (k-g-j) [reply packages/snap/src/utils/validation.ts]: Addressed here [60ee50b](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/60ee50be21e3477c78a5b456fb7ba3c3ec5ab13c)

File: packages/snap/src/keyring.ts (hunk 4 above)
Comment (k-g-j) [reply packages/snap/src/keyring.ts]: Addressed here [703efa7](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/703efa74adad3f09bba8f3a7a0cc0b037b58d1b2)

File: packages/snap/src/keyring.test.ts [hunk 9]
```diff
@@ -121,26 +121,45 @@ describe('Keyring', () => {
           invalidConfig.simpleAccountFactory as string
         }`,
       );
-      invalidConfig.simpleAccountFactory = config.simpleAccountFactory!;
-      invalidConfig.entryPoint = '0xNotAnAddress';
+    });
+    it('should not set the config with and invalid entryPoint address', async () => {
```
[the remark this answers is not in the document]
Comment (k-g-j) [reply packages/snap/src/keyring.test.ts]: Addressed here [301fb96](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/301fb96252f28b38ace4f52ee14b16daa2c6a101)

File: packages/snap/src/keyring.test.ts (hunk 6 above)
Comment (k-g-j) [reply packages/snap/src/keyring.test.ts]: Addressed here [301fb96](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/301fb96252f28b38ace4f52ee14b16daa2c6a101)

File: packages/snap/src/utils/validation.ts (hunk 5 above)
Comment (k-g-j) [reply packages/snap/src/utils/validation.ts]: Addressed here [39e96b5](https://github.com/MetaMask/snap-account-abstraction-keyring/pull/20/commits/39e96b574777574fd43804129565fcdea3be01a9)

**Review by montelaidev — APPROVED** [review]

**Review by montelaidev — APPROVED** [review]

**Pull request merged**